### PR TITLE
i#2283: Adds umbra_clear_redundant_blocks 

### PR DESCRIPTION
**Pull request by johnfxgalea**
Adds a new API function which allows the user to trigger the removal of redundant blocks that consist of only default values.

It is useful when low on memory.

Issue #2283

### CODE REVIEW

File: umbra/umbra_32.c [hunk 1]
```diff
@@ -1049,3 +1049,42 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
...
+{
+    byte *shadow_data;
+    uint i, j;
+
+    if (map == NULL)
+        return DRMF_ERROR_INVALID_PARAMETER;
+
+    /* Umbra needs to have create-on-touch optimization enabled. */
+    if (!TEST(UMBRA_MAP_CREATE_SHADOW_ON_TOUCH, map->options.flags))
+        return DRMF_ERROR;
```
Comment (derekbruening) [inline umbra/umbra_32.c]: Is there a more specific error that would fit: DRMF_ERROR_FEATURE_NOT_AVAILABLE or DRMF_ERROR_INVALID_PARAMETER?

File: umbra/umbra_64.c [hunk 2]
```diff
@@ -1369,3 +1369,9 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
     }
     return false;
 }
+
+drmf_status_t
+umbra_clear_redundant_blocks(umbra_map_t *map) {
```
Comment (derekbruening) [inline umbra/umbra_64.c]: style: { on own line

File: umbra/umbra_64.c [hunk 3]
```diff
@@ -1369,3 +1369,9 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
...
+    /* No operation needed for x64. */
+    return DRMF_ERROR_FEATURE_NOT_AVAILABLE;
+}
```
Comment (derekbruening) [inline umbra/umbra_64.c]: Waiting for a test as well?

File: umbra/umbra_64.c (hunk 2 above)
Comment (johnfxgalea) [reply umbra/umbra_64.c]: Ah sorry about this. I have another coding style for my other projects and keep on missing such issues. When I applied clang-format on Umbra code, it modified the entire file - hence me avoiding formatting.

Comment (derekbruening) [reply umbra/umbra_64.c]: I have #2107 on my list.

**Comment by johnfxgalea** [comment]
Failures due to #2180

File: umbra/umbra_64.c (hunk 2 above)
Comment (johnfxgalea) [reply umbra/umbra_64.c:1374]: Done.

File: umbra/umbra_32.c (hunk 1 above)
Comment (johnfxgalea) [reply umbra/umbra_32.c]: Done.

**Comment by johnfxgalea** [comment]
@derekbruening Please take a look when you get a chance. Thanks!

File: umbra/umbra_64.c (hunk 3 above)
Comment (johnfxgalea) [reply umbra/umbra_64.c:1378]: I added a test.

**Comment by johnfxgalea** [comment]
Appveyor failed due to #2178.

**Comment by derekbruening** [comment]
> @derekbruening Please take a look when you get a chance. Thanks!

Sorry missed this.  Will try to look soon.

**Comment by johnfxgalea** [comment]
> > @derekbruening Please take a look when you get a chance. Thanks!
> 
> Sorry missed this. Will try to look soon.

DW, no rush.

File: umbra/umbra_32.c [hunk 4]
```diff
@@ -1049,3 +1049,53 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
...
+    uint i, j;
+
+    if (map == NULL)
+        return DRMF_ERROR_INVALID_PARAMETER;
+
+    if (count != NULL)
+        *count = 0;
+
+    /* Umbra needs to have create-on-touch optimization enabled. */
+    if (!TEST(UMBRA_MAP_CREATE_SHADOW_ON_TOUCH, map->options.flags)){
```
Comment (derekbruening) [inline umbra/umbra_32.c:1066]: I think the doxygen doc should state that it will fail w/o this option.

File: umbra/umbra_32.c [hunk 5]
```diff
@@ -1049,3 +1049,53 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
...
+        return DRMF_ERROR_INVALID_PARAMETER;
+    }
+
+    umbra_map_lock(map);
+    for (i = 0; i < SHADOW_TABLE_ENTRIES; i++) {
+        shadow_data = shadow_table_get_block(map, i);
+        /* Redundant blocks must be "normal". */
+        if (shadow_table_is_in_normal_block(map, shadow_data)) {
```
Comment (derekbruening) [inline umbra/umbra_32.c:1074]: suggest: invert and continue to avoid indentation below

File: umbra/umbra_32.c [hunk 6]
```diff
@@ -1049,3 +1049,53 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
...
+            /* Check whether the entire block consists of default values. */
+            bool is_all_default_val = true;
+            for (j = 0; j < map->shadow_block_size; j++){
+                if (shadow_data[j] != (byte) map->options.default_value) {
+                    is_all_default_val = false;
+                    break;
+                }
+            }
+            /* Delete block only if it consists of default values. */
+            if (is_all_default_val) {
```
Comment (derekbruening) [inline umbra/umbra_32.c:1084]: suggest: invert and continue

Comment (johnfxgalea) [reply umbra/umbra_32.c:1084]: Done.

File: umbra/umbra_32.c (hunk 5 above)
Comment (johnfxgalea) [reply umbra/umbra_32.c:1074]: Done.

File: umbra/umbra_32.c (hunk 4 above)
Comment (johnfxgalea) [reply umbra/umbra_32.c:1066]: Done.

**Comment by johnfxgalea** [comment]
@derekbruening, I applied clang-format to the umbra files I modified to facilitate my future additional PRs.

**Pull request merged**